### ISSUE TITLE
Make livereload usable for custom rules 

### ISSUE THREAD

**Issue by Ubehebe** [bazelbuild/bazel-watcher#248]
First, thanks for your work on ibazel. It's an important part of my development workflow, as I discuss below.

I'd like to start a conversation about making ibazel's livereload feature easier to integrate with custom rules. Here's the current situation. If I have an executable Starlark rule foo, `ibazel run :some_foo_target` works OK without any modification to the foo rule implementation. When any of the target's inputs change, ibazel kills whatever the foo target was doing and restarts it with the fresh files. For executable rules that do something like print to a terminal, this is an ideal development experience.

For rules that do something more expensive/complex like bring up a web server and open a browser, the base ibazel behavior is not a good development experience. (For example, in a webapp rule I've written, editing a CSS or JS file causes ibazel to kill both the server and the browser, then bring them back a few seconds later.) It would be great if there was a simple API for custom rules to gain livereload superpowers.

My vision is this: **any rule that produces JS, CSS, or HTML (or any outputs that it is valuable to show in a browser) should be able to be invoked with ibazel run and have livereload just work**.

Currently, it is possible to get a custom rule to work with livereload, but it isn't easy. The rule has to set up a web server that can read the magic `IBAZEL_LIVERELOAD_URL` environment variable on startup and inject script tags pointing to that URL in the HTML files it serves. ts_devserver does this, but it also does a lot of other stuff (concatenating JS, injecting require.js) that may not be appropriate for custom rules.

I think there is an opportunity here to make ibazel useful for a great deal of frontend toolchains by providing this logic. I've implemented it in a private repository, and would be willing to do the work to upstream it here. But before talking about technical details, I wanted to get your overall thoughts.

Maintainers, what do you think? Does this fit in with the scope and purpose of this project? (I know there's some overlap between ibazel and ts_devserver, and they seem to have been developed tightly coupled. I'm asking here first because the ibazel repo seems like the right level of generality for this logic.)

**Comment by achew22** [comment]
Thanks for the kind words!

> Maintainers, what do you think? Does this fit in with the scope and purpose of this project? (I know there's some overlap between ibazel and ts_devserver, and they seem to have been developed tightly coupled. I'm asking here first because the ibazel repo seems like the right level of generality for this logic.)

It definitely sounds interesting. My concerns are going to be around the combinatorial explosion of languages that might need to be supported. Maybe as a first step you could make the private repo public or share it with me so I can figure out what the code you're proposing looks like. I'm definitely amenable to moving the logic in here as long as we find a reasonable abstraction layer.

**Comment by Ubehebe** [comment]
The API I have in mind would be completely language- and rule-agnostic. Briefly, if you are the author of a rule foo with an implementation like this:

```python
def foo_impl(ctx):
  outputs = ... # generate JS/CSS/HTML/etc. somehow
  return [
    DefaultInfo(files = outputs),
    # more providers we don't need to know about
  ]
```

then we would provide a `serve` function that returns a Runfiles object that you can just drop into the runfiles returned by the rule:

```python
def foo_impl(ctx):
  outputs = ... # generate JS/CSS/HTML/etc. somehow
  return [
    DefaultInfo(
      files = outputs,
      runfiles = serve(outputs), # or choose some subset to serve
    ), 
  ]
```

(`serve` would also write to `ctx.outputs.executable`.) This 1-line change would make your rule now work with ibazel livereload.

> Maybe as a first step you could make the private repo public or share it with me so I can figure out what the code you're proposing looks like.

My current implementation is in Java with a bunch of Dagger bindings. It's probably more complex than it needs to be, so over the next few days I'll try to get a simple Go implementation working and publish it.

**Comment by achew22** [comment]
I'm definitely intrigued by this proposal, but I'm not sure I have enough context on it to fully understand.

1. `serve` will be passed the `ctx` object as well, right? If not, how does it write to the outputs?
1. Is `serve` is going to put an additional file in the runfiles directory? Does this mean a `bazel query` on the target to get all its configuration settings out?

**Comment by Ubehebe** [comment]
I've published [bazel-runfiles-server](https://github.com/Ubehebe/bazel-runfiles-server) as a proof of concept. It defines the server (still in Java instead of Go, but pretty simple), as well Starlark APIs for using the server from custom rules with basically a 1-line change.

The repo includes examples of using this from a rule under your control (I wrote a simple rule to render TeX to HTML using KaTeX, and got livereload working), as well as how to get livereload working if you don't control the rule (I used the third-party [Pandoc rules](https://github.com/ProdriveTechnologies/bazel-pandoc) to render Markdown to HTML, and defined a separate ibazel-aware serve() target). The repo's readme discusses these use cases in more detail.

Regarding your specific questions, yes, `serve_this` takes `ctx` and returns a runfiles object, which rule authors are expected to propagate as their own runfiles. This means that the rule cannot already be executable (since `serve_this` writes to `ctx.outputs.executable`). I'm not sure 

> Is serve is going to put an additional file in the runfiles directory? Does this mean a bazel query on the target to get all its configuration settings out?

Yes, serve_this adds additional runfiles. But I'm not sure why this requires a bazel query; I got all this working without modifying ibazel.

I'll let you look over the repo for a few days. Hope this clarifies the scope of the project. Thanks for your time!

**Comment by Ubehebe** [comment]
Gentle ping for feedback on this proposal. Do you think bazel-runfiles-server would be appropriate for upstreaming here? I got in touch with a rules_pandoc committer about this and they responded enthusiastically, as you can see in the link above.

**Comment by achew22** [comment]
So sorry to be slow on this. I think this does a good job of showing where you would like things to be.

First, this is a very cool idea. The thing I like the most is the ability to turn any other rule into a [live reloading target](https://github.com/Ubehebe/bazel-runfiles-server/blob/master/examples/BUILD#L26). Way to think creatively and bring it to fruition. This is almost certainly something I will use personally if we merge it. My question to you is, how should we go about merging it? What pieces would you like to live in iBazel's repo and what would you like to have external? My personal preference is to stick with a single language for the project, would you be receptive to the server being rewritten in `golang` as a part of iBazel's repo?

**Comment by Ubehebe** [comment]
Thanks for the feedback!

The main risk I see with upstreaming this is that people are going to want more complexity put into the server: CORS headers, certificates, dynamic logic, etc. I like the simplicity of "serve whatever is in runfiles with zero extra semantics", so hopefully we can stay close to that.

>  would you be receptive to the server being rewritten in golang as a part of iBazel's repo?

That's a reasonable request. I'll port it to Go over the next couple of weeks and ping this thread when it's ready. By then, I should have a more precise idea of how to land the code in this repo.

**Comment by Ubehebe** [comment]
Update: I've [ported](https://github.com/Ubehebe/bazel-runfiles-server/commit/3c75ca5f39c9dce1abdf877678125cbe6b27998b) the repo to Go. Thanks for your patience. I'm going to rebase the repo against upstream, and I should have a PR ready by the end of the week.

**Comment by joneshf** [comment]
Can we re-visit this issue? Trying to get livereload to work with `ibazel` is still non-trivial. The suggested ideas here and the proof of concept seem like they'd be incredibly easier to use.

**Comment by achew22** [comment]
There isn't anything that prevents you from using the poc. Nothing in ibazel was changed so there isn't any code modification necessary in ibazel to use it.

**Comment by joneshf** [comment]
Right, that's what I ended up doing. The part I'd like to revisit is merging that work into this repo.

The reason I'd like to revisit merging into this repo is three-fold:
1. As it sits now, it's pretty hard to get livereload working. I couldn't get any of the other suggestions in this repo to really work. The suggestions of using environment variables doesn't reliably work since different servers that do some kind of livereload don't always look at environment variables. It's also hard because not every server follows the livereload protocol, so it's kind of hit or miss whether it's going to work.
1. There's already a PR that seems to only have been closed out of staleness: #261.
2. There's no telling how long that proof-of-concept repo is going to be around. The alternative is for people to fork that repo, vendor the rules in each workspace, or something else. That works, it's just more friction to having livereload working.

If there's a different alternative you'd prefer, that's fine too. I don't particularly care about how it's implemented. I'm not vested in seeing that proof-of-concept proliferating. It's an incredibly interesting application of all the pieces, but all I'm hoping is that there's some (easy) way to do livereload with `ibazel`.
